bftengine/tests/testViewChange/testViewChange.cpp

```diff
@@ -165,73 +138,182 @@ TEST(testViewchangeSafetyLogic_test, computeRestrictions) {
+                                              (uint64_t)1234567,
+                                              kRequestLength,
+                                              (const char*)requestBuffer2,
+                                              (uint64_t)1000000);
+
+  // Generate the PrePrepare from the primary for the clientRequest2
+  auto* ppMsg2 = new PrePrepareMsg(pRepInfo->primaryOfView(curView2),
+                                   curView2,
+                                   assignedSeqNum,
+                                   bftEngine::impl::CommitPath::SLOW,
+                                   false,
+                                   clientRequest2->size());
+  ppMsg2->addRequest(clientRequest2->body(), clientRequest2->size());
+  ppMsg2->finishAddingRequests();
+  // Here we generate a valid Prepare Certificate for clientRequest2 with
+  // View=1, this is going to be the Prepare Certificate that has to be
+  // considered for the sequence number -> assignedSeqNum, because it will
+  // be the one with the highest View for assignedSeqNum.
+  PrepareFullMsg* pfMsg2 =
+      PrepareFullMsg::create(curView2, assignedSeqNum, pRepInfo->primaryOfView(curView2), buff, sizeof(buff));
+
+  ViewChangeMsg** viewChangeMsgs = new ViewChangeMsg*[N];
+
+  viewChangeMsgs[0] = new ViewChangeMsg(0, curView2, lastStableSeqNum);
+  viewChangeMsgs[1] = nullptr;
+  viewChangeMsgs[2] = new ViewChangeMsg(2, curView2, lastStableSeqNum);
+  viewChangeMsgs[3] = new ViewChangeMsg(3, curView2, lastStableSeqNum);
+
+  // Add the first Prepare Certificate which we expect to be ignored
+  // to the View Change Msg from Replica 2
+  viewChangeMsgs[2]->addElement(*pRepInfo,
+                                assignedSeqNum,
+                                ppMsg1->digestOfRequests(),
+                                ppMsg1->viewNumber(),
+                                true,
+                                ppMsg1->viewNumber(),
+                                pfMsg1->signatureLen(),
+                                pfMsg1->signatureBody());
+
+  // Add the second Prepare Certificate which we expect to be considered in the next view
+  // to the View Change Msg from Replica 3
+  viewChangeMsgs[3]->addElement(*pRepInfo,
+                                assignedSeqNum,
+                                ppMsg2->digestOfRequests(),
+                                ppMsg2->viewNumber(),
+                                true,
+                                ppMsg2->viewNumber(),
+                                pfMsg2->signatureLen(),
+                                pfMsg2->signatureBody());
+
+  auto VCS = ViewChangeSafetyLogic(N, F, C, &dummyVerifier_, PrePrepareMsg::digestOfNullPrePrepareMsg());
 
   SeqNum min{}, max{};
   VCS.computeRestrictions(viewChangeMsgs, 0, min, max, restrictions);
 
   for (int i = 0; i < kWorkWindowSize; i++) {
     if (i == assignedSeqNum - 1) {
```

Comment:
why `assignedSeqNum - 1` instead of `assignedSeqNum` ?

Reply:
My mistake, in computeRestrictions() the second argument is the stable sequence number extracted from the View Change messages. This function returns "min" and "max" sequence numbers in the current working window, so basically "i == assignedSeqNum - min" should be the check.